src/transactionsOutcomeParsers/tokenManagementTransactionsOutcomeParser.ts

```diff
@@ -6,57 +6,53 @@ import { bufferToBigInt } from "../smartcontracts/codec/utils";
+    parseIssueSemiFungible(transactionOutcome: TransactionOutcome): { tokenIdentifier: string }[] {
         this.ensureNoError(transactionOutcome.transactionLogs.events);
 
-        const event = this.findSingleEventByIdentifier(transactionOutcome, "issueSemiFungible");
-        const identifier = this.extractTokenIdentifier(event);
-
-        return { tokenIdentifier: identifier };
+        const events = this.findEventsByIdentifier(transactionOutcome, "issueSemiFungible");
+        return events.map((event) => ({ tokenIdentifier: this.extractTokenIdentifier(event) }));
     }
 
-    parseRegisterMetaEsdt(transactionOutcome: TransactionOutcome): { tokenIdentifier: string } {
+    parseRegisterMetaEsdt(transactionOutcome: TransactionOutcome): { tokenIdentifier: string }[] {
         this.ensureNoError(transactionOutcome.transactionLogs.events);
 
-        const event = this.findSingleEventByIdentifier(transactionOutcome, "registerMetaESDT");
-        const identifier = this.extractTokenIdentifier(event);
-
-        return { tokenIdentifier: identifier };
+        const events = this.findEventsByIdentifier(transactionOutcome, "registerMetaESDT");
+        return events.map((event) => ({ tokenIdentifier: this.extractTokenIdentifier(event) }));
     }
 
-    parseRegisterAndSetAllRoles(transactionOutcome: TransactionOutcome): { tokenIdentifier: string; roles: string[] } {
+    parseRegisterAndSetAllRoles(
+        transactionOutcome: TransactionOutcome,
+    ): { tokenIdentifier: string; roles: string[] }[] {
         this.ensureNoError(transactionOutcome.transactionLogs.events);
+        const registerEvents = this.findEventsByIdentifier(transactionOutcome, "registerAndSetAllRoles");
+        const setRoleEvents = this.findEventsByIdentifier(transactionOutcome, "ESDTSetRole");
 
-        const registerEvent = this.findSingleEventByIdentifier(transactionOutcome, "registerAndSetAllRoles");
-        const tokenIdentifier = this.extractTokenIdentifier(registerEvent);
-
-        const setRoleEvent = this.findSingleEventByIdentifier(transactionOutcome, "ESDTSetRole");
-        const encodedRoles = setRoleEvent.topics.slice(3);
-
-        let roles: string[] = [];
-        for (const role of encodedRoles) {
-            roles = roles.concat(this.decodeTopicAsString(role));
+        if (registerEvents.length !== setRoleEvents.length) {
+            throw new ErrParseTransactionOutcome(
+                "Register Events and Set Role events mismatch. Should have the same number of events.",
+            );
         }
 
-        return { tokenIdentifier: tokenIdentifier, roles: roles };
+        return registerEvents.map((registerEvent, index) => {
+            const tokenIdentifier = this.extractTokenIdentifier(registerEvent);
+            const encodedRoles = setRoleEvents[index].topics.slice(3);
+            const roles = encodedRoles.map((role) => this.decodeTopicAsString(role));
+            return { tokenIdentifier, roles };
+        });
```

Comment:
Could it be possible (in real-life) that the order of `registerEvents` is not correlated with the order of `setRoleEvents`?

Reply:
Hopefully not